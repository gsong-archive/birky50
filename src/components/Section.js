import React from "react";

import { css } from "emotion";

import { SectionProvider } from "../contexts/SectionContext";

const forwardRef = (render, displayName = "") => {
  render.displayName = displayName;
  return React.forwardRef(render);
};

export default forwardRef(
  (props, ref) => (
    <SectionProvider id={props.id}>
      <section
        {...props}
        ref={ref}
        className={css`
          ${props.className};
          padding: 1.5rem;
        `}
      >
        {props.children}
      </section>
    </SectionProvider>
  ),
  "Section"
);
