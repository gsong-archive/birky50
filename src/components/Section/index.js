//@flow
import React from "react";

import { css } from "emotion";

import { SectionProvider } from "../../contexts/SectionContext";

const forwardRef = (render, displayName = "") => {
  render.displayName = displayName;
  // $FlowFixMe
  return React.forwardRef(render);
};

export default forwardRef((props, ref) => {
  const sectionProps = Object.assign({}, props, {
    ref,
    className: css`
      ${props.className};
      padding: 16px 24px 24px;
    `,
  });
  delete sectionProps.tag;
  const section = React.createElement(props.tag, sectionProps);

  return <SectionProvider id={props.id}>{section}</SectionProvider>;
}, "Section");
