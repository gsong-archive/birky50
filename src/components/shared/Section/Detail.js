import styled, { css } from "react-emotion";

export default styled.div(
  css`
    padding: 0 24px;
  `
);
