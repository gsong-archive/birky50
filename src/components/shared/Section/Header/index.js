//@flow strict
import * as React from "react";

import facepaint from "facepaint";
import styled, { css } from "react-emotion";

import SectionContext from "../../../../contexts/SectionContext";
import { calcSpacing, convertListToEm } from "../../../../styles/utils";

type Props = { color: string, children: React.Node };

export default ({ color, children }: Props) => (
  <SectionContext.Consumer>
    {id => (
      <Container
        className={css`
          border-top: 8px solid ${color};
        `}
      >
        <H1 id={`${id}-description`}>{children}</H1>
      </Container>
    )}
  </SectionContext.Consumer>
);

const breakpoints = convertListToEm([350]);
const mq = facepaint(breakpoints.map(bp => `@media (max-width: ${bp}em)`));

const calcPaddingTopPx = (totalRows, fontSize) =>
  `${(calcSpacing(totalRows, fontSize) / 2) * 16}px`;

const Container = styled.div(
  css`
    position: sticky;
    top: 0;
    background-color: hsla(0, 0%, 100%, 0.8);
  `,
  mq({
    padding: [
      `${calcPaddingTopPx(3, 1.75)} 24px`,
      `${calcPaddingTopPx(2.5, 1.5)} 24px`,
    ],
  })
);

const H1 = styled.h1(
  css`
    font-weight: normal;
    margin: 0 0;
  `,
  mq({ fontSize: ["1.75rem", "1.5rem"] })
);
