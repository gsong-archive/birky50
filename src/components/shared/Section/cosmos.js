//@flow
import React from "react";

import Section from "./index";
import { WindowSizeProvider } from "../../../contexts/WindowSizeContext";

import type { Props } from "./index";

export default (props: Props) => (
  <WindowSizeProvider>
    <Section {...props} />
  </WindowSizeProvider>
);
