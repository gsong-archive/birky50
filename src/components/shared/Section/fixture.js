//@flow
import React from "react";

import { createFixture } from "react-cosmos-flow/fixture";

import Section from "./cosmos";
import sections from "../../../data/sections";

export default sections.map(section => {
  const ref = React.createRef();

  return createFixture({
    name: section.navLabel,
    description:
      "`data` prop is a section spec, `forwardedRef` is a React component ref",
    component: Section,
    props: {
      data: section,
      forwardedRef: ref,
    },
  });
});
