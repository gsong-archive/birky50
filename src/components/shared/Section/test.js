//@flow
import React from "react";

import { render } from "react-testing-library";

import Section from "./index";
import sections from "../../../data/sections";

test("Section is rendered", () => {
  const ref = React.createRef();
  const { container } = render(
    <Section data={sections[1]} forwardedRef={ref} />
  );

  expect(container.firstChild).toMatchSnapshot();
});
