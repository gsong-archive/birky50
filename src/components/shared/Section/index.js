//@flow
import * as React from "react";

import { css } from "emotion";

import Container from "./Container";
import Header from "./Header";

import type { SectionDatum } from "../../Section.type";

export type Props = {
  data: SectionDatum,
  forwardedRef: { current: null | React$ElementRef<React.ElementType> },
};

export default ({ data, forwardedRef }: Props) => {
  const {
    id,
    sectionLabel,
    sectionTag,
    SectionComponent,
    LabelComponent,
    color,
  } = data;
  // flowlint-next-line sketchy-null-string:off
  const tag = sectionTag ? sectionTag : "section";

  return (
    <Container
      tag={tag}
      id={id}
      key={id}
      ref={forwardedRef}
      aria-labelledby={`${id}-description`}
    >
      <Header color={color}>
        <LabelComponent label={sectionLabel} />
      </Header>

      <div
        className={css`
          padding: 0 24px;
        `}
      >
        <SectionComponent />
      </div>
    </Container>
  );
};
