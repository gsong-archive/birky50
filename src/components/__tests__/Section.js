import React from "react";

import { render } from "react-testing-library";

import Section from "../Section";

test("Section renders properly", () => {
  const ref = React.createRef();
  const { container } = render(
    <Section id="id" ref={ref}>
      Child
    </Section>
  );

  expect(container.firstChild).toMatchSnapshot();
});
