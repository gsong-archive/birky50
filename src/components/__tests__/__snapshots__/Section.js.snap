// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section renders properly 1`] = `
.emotion-0 {
  padding: 1.5rem;
}

<section
  id="id"
  class="emotion-0"
>
  Child
</section>
`;
